libbeat/paths/paths.go

```diff
@@ -0,0 +1,143 @@
+//
+// cfgfilePath := paths.Resolve(paths.Config, "beat.yml"
+package paths
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+var (
+	homePath   = flag.String("path.home", "", "Home path")
+	configPath = flag.String("path.config", "", "Configuration path")
+	dataPath   = flag.String("path.data", "", "Data path")
+)
+
+type Path struct {
+	Home   string
+	Config string
+	Data   string
+}
+
+// FileType is an enumeration type representing the file types.
+// Currently existing file types are: Home, Config, Data
+type FileType string
+
+const (
+	Home   FileType = "home"
+	Config FileType = "config"
+	Data   FileType = "data"
+)
+
+var paths Path
+
+// InitPaths sets the default paths in the configuration based on CLI flags,
+// configuration file and default values. It also tries to create the data
+// path with mode 0755 and returns an error on failure.
+func InitPaths(cfg *Path) error {
+	err := initPaths(cfg)
+	if err != nil {
+		return err
+	}
+
+	// make sure the data path exists
+	err = os.MkdirAll(paths.Data, 0755)
+	if err != nil {
+		return fmt.Errorf("Failed to create data path %s: %v", paths.Data, err)
+	}
+
+	return nil
+}
+
+// initPaths sets the default paths in the configuration based on CLI flags,
+// configuration file and default values.
+func initPaths(cfg *Path) error {
+	// overwrite paths from CLI flags
+	if homePath != nil && len(*homePath) > 0 {
+		cfg.Home = *homePath
+	}
+	if configPath != nil && len(*configPath) > 0 {
+		cfg.Config = *configPath
+	}
+	if dataPath != nil && len(*dataPath) > 0 {
+		cfg.Data = *dataPath
+	}
+
+	// default for the home path is the binary location
+	if len(cfg.Home) == 0 {
+		var err error
+		cfg.Home, err = filepath.Abs(filepath.Dir(os.Args[0]))
+		if err != nil {
+			return fmt.Errorf("The absolute path to %s could not be obtained. %v",
+				os.Args[0], err)
+		}
+	}
+
+	// default for config path
+	if len(cfg.Config) == 0 {
+		cfg.Config = cfg.Home
+	}
+
+	// default for data path
+	if len(cfg.Data) == 0 {
+		cfg.Data = filepath.Join(cfg.Home, "data")
+	}
+
+	paths = *cfg
+
+	return nil
+}
+
+// Resolve resolves a path to a location in one of the default
+// folders. For example, Resolve(Home, "test") returns an absolute
+// path for "test" in the home path.
+func Resolve(fileType FileType, path string) string {
+	// absolute fpaths are not changed
+	if filepath.IsAbs(path) {
+		return path
+	}
+
+	switch fileType {
+	case Home:
+		return filepath.Join(paths.Home, path)
+	case Config:
+		return filepath.Join(paths.Config, path)
+	case Data:
+		return filepath.Join(paths.Data, path)
+	default:
+		logp.WTF("Unknown file type: %s", fileType)
+		return ""
+	}
+}
+
+// LogPaths uses logp to print the configured paths at Info level.
+// This is a separate method because it needs to be called after
+// logging was initialized. InitPaths is typically called before
+// logging is initialized.
+func LogPaths() {
```

Comment:
Consider decoupling this method from the logger. Instead have a method that returns a string representation of the `Path` then allow the caller to log it.